### PR TITLE
CheriBSD main

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

File: pycheribuild/projects/cross/gdb.py [hunk 1]
```diff
@@ -240,7 +240,10 @@ class BuildGDB(BuildGDBBase):
     repository = GitRepository(
         "https://github.com/CTSRD-CHERI/gdb.git",
         # Branch name is changed for every major GDB release:
-        default_branch="mips_cheri-8.3", force_branch=True,
+        default_branch="mips_cheri-8.3",
+        old_branches={"mips_cheri_7.12": "mips_cheri-8.3",
```
Comment (bsdjhb) [inline pycheribuild/projects/cross/gdb.py:244]: It would be handy to use self.default_branch as the value in the dictionary perhaps.

Comment (jrtc27) [reply pycheribuild/projects/cross/gdb.py:244]: I guess you could assign special meaning to None as self.default_branch

File: pycheribuild/projects/project.py [hunk 2]
```diff
@@ -1770,24 +1780,22 @@ def update(self, current_project: "Project", *, src_dir: Path, base_project_sour
...
-            if status.stdout.startswith(b"## ") and not status.stdout.startswith(
-                    b"## " + default_branch.encode("utf-8") + b"..."):
-                current_branch = status.stdout[3:status.stdout.find(b"...")].strip()
-                current_project.warning("You are trying to build the", current_branch.decode("utf-8"),
+        if src_dir.exists() and (self.force_branch or self.old_branches):
+            if self.force_branch:
+                assert self.old_branches is None, "cannot set both force_branch and old_branches"
+                default_branch = self.get_default_branch(current_project, include_per_target=True)
+                assert default_branch, "default_branch must be set if force_branch is true!"
+            current_branch = self.get_current_branch(src_dir).decode("utf-8")
```
Comment (jrtc27) [inline pycheribuild/projects/project.py:1788]: It deviates more from the original code, but maybe hoist this above the force_branch case? Then you have the two cases setting default_branch grouped together

Comment (bsdjhb) [reply pycheribuild/projects/project.py:1788]: Yeah, I almost did that and honestly prefer it that way as well.

File: pycheribuild/projects/project.py [hunk 3]
```diff
@@ -1770,24 +1780,22 @@ def update(self, current_project: "Project", *, src_dir: Path, base_project_sour
...
+            if self.old_branches:
+                default_branch = self.old_branches.get(current_branch)
```
Comment (jrtc27) [inline pycheribuild/projects/project.py:1790]: This won't work if current_branch isn't in old_branches; it won't raise an exception like `[]`, instead it'll give None and blow up later. Easiest solution is to add current_branch as the optional default value argument.

File: pycheribuild/projects/cross/gdb.py (hunk 1 above)
Comment (bsdjhb) [reply pycheribuild/projects/cross/gdb.py:244]: Hmm, that's an interesting idea, but that seems kind of odd?  It also makes it a bit more of a pain to deal with in the implementation.  The other option I've thought about for the gdb case is just using a local default_branch variable in this function.

File: pycheribuild/projects/project.py (hunk 3 above)
Comment (jrtc27) [reply pycheribuild/projects/project.py:1790]: Oh, never mind, you already check for None... though if you want to have `"foo": None` mean `self._default_branch` then you do need to be able to distinguish the cases

Comment (bsdjhb) [reply pycheribuild/projects/project.py:1790]: I check for None below (I purposely didn't want an exception)

Comment (bsdjhb) [reply pycheribuild/projects/project.py:1790]: Yeah, this is why using None to mean default branch is a bit messier.

File: pycheribuild/projects/cross/gdb.py (hunk 1 above)
Comment (jrtc27) [reply pycheribuild/projects/cross/gdb.py:244]: Yeah that's less gross

**Review by jrtc27 — APPROVED** [review]
(with the obvious caveat of "don't land the final commit until we're ready"; others can go in earlier so it's just the CheriBSD change pending)

**Comment by jrtc27** [comment]
(GitHub doesn't update the diff / list of commits unless you change the base...)

**Pull request merged**